### PR TITLE
[JUJU-2519] Remove groups on start instance failure

### PR DESCRIPTION
**Pull request by SimonRichardson**
If we're unable to provision an instance during startup, then ensure that we clean up any groups we may have created during the process.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

Setup microstack (can use any OpenStack), set secgroup quota.

```sh
$ juju bootstrap --bootstrap-series=focal \
   --metadata-source=~/simplestreams \
   --model-default network=test \
   --model-default external-network=external \
   --bootstrap-constraints allocate-public-ip=true \
   microstack microstack
$ juju add-machine -n 1
$ juju remove-machine 0
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1940637

### CODE REVIEW

**Comment by hmlanigan** [comment]
having issues with qa - microstack doesn't fails init for me and i can't edit quotas on serverstack.

trying to hit my server quota now.

**Review by hmlanigan — APPROVED** [review]
looks okay - however i'm having qa difficulties

File: provider/openstack/provider.go [hunk 1]
```diff
@@ -2125,6 +2143,14 @@ func (e *Environ) terminateInstances(ctx context.ProviderCallContext, ids []inst
 	var firstErr error
 	novaClient := e.nova()
 	for _, id := range ids {
+		// Attempt to destroy any groups that could have been created when
+		// creating the instance.
+		if err := e.terminateInstanceSecurityGroups(ctx, id); err != nil {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply provider/openstack/provider.go:2148]: You're right, let me attempt to clean this up in a better way.

File: provider/openstack/provider.go [hunk 2]
```diff
@@ -1185,12 +1184,24 @@ func (e *Environ) startInstance(
 		}
 	}
 
+	// Clean up any groups that we have created if we fail to start the
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply provider/openstack/provider.go:1187]: Sure.

**Comment by SimonRichardson** [comment]
@hmlanigan I cleaned the security group clean up so it correctly removes them during the termination process for all instances.

**Comment by hmlanigan** [comment]
added do not merge - so I have a chance to re-review as requested

**Review by hmlanigan — APPROVED** [review]
LGTM - tested with network port security enable and disabled. Unable to test error cases - no ability to change quota.

**Pull request merged**